Use mutable var instead of statet in model

abailly-iohk opened the pull request:
Replace `StateT (Nodes m)` with `ReaderT (TVar m (Nodes m))` in the `RunModel` so that we can make `RunModel m` a `MonadThrow` and therefore use proper `RunException` exceptions when something goes wrong on the SUT during property checks, replacing crude `error ...` calls with better formed exceptions.

abailly committed:
Use exception in Model to signal unexpected SUT behaviour

Exception do not play well with StateT which was an oversight in the
initial design of the `RunModel`. By using `ReaderT` and proper
exception type, we should have better failure reports: io-sim will
catch  those exceptions and therefore dump the logs and execution
trace, whereas now (AFAIK) the error is caught outside of io-sim and
we don't really have much information about the context.

abailly committed:
Actually use the correct type to print the trace

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -366,14 +371,23 @@ data Nodes m = Nodes
+    ReaderT $ \_ -> do
       a <- r
-      pure (a, s)
+      pure a
```
I think a `lift = RunMonad . lift` would yield the same semantics, no?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -474,11 +488,17 @@ seedWorld seedKeys seedCP = do
+  s <- ask
+  lift $
+    atomically $
+      modifyTVar (nodesState s) $ \n ->
```
Can we define a `modifyRunState :: MonadSTM m => (Nodes m -> Nodes m) -> m` instead? It would hide the fact that we are actually using a `ReaderT TVar` under the hood and be easier to update in case this changes further.

ch1bo replied:
or an `-> STM` variant where you would have control over the atomically section.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -474,11 +488,17 @@ seedWorld seedKeys seedCP = do
+          }
+
+gets :: MonadSTM m => (Nodes m -> a) -> RunMonad m a
+gets f = f <$> (ask >>= lift . readTVarIO . nodesState)
```
Seeing my comment above and this definition.

Could: define `instance MonadSTM m => MonadState (RunMonad m)`

This would yield appropriate definitions for `gets` and `modify` (besides `get` and `put`)

ch1bo reviewed (approved):
Some further refactoring into `MonadState` possible and/or hiding the `ReaderT` from actual usage sites.

abailly committed:
Address review comments

Make RunMonad a MonadState (Nodes m)  which under the hood uses the
TVar so that code can use get/put/modify interface.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -561,16 +577,16 @@ performNewTx party tx = do
+performAbort :: (MonadDelay m, MonadSTM m, MonadThrow m) => Party -> RunMonad m ()
 performAbort party = do
-  Nodes{nodes} <- get
+  nodes <- gets nodes
```
I liked the previous better. Also reading this on github makes me wonder.. is this not a naming conflict?
